### PR TITLE
refactor run; move to stage.utils

### PR DESCRIPTION
**Pull request by skshetry**
Pre-requisite PR for #5253. Moving some code to `stage.utils` to reduce duplications, that might be there when we introduce `stage add/create`.

Also changed `loads_params_from_cli` and `parse_params_from_cli` to `loads_params` and `parse_params` respectively.

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/stage/__init__.py [hunk 1]
```diff
@@ -666,6 +669,9 @@ def merge(self, ancestor, other):
 
         self.outs[0].merge(ancestor_out, other.outs[0])
 
+    def dump(self, update_lock: bool = False):
```
Comment (skshetry) [inline dvc/stage/__init__.py:672]: Added a `dump` method on `Stage` itself.

File: dvc/dvcfile.py [hunk 2]
```diff
@@ -381,12 +386,19 @@ def merge(self, ancestor, other):
...
-        # fallback to single stage file for better error messages
-        return SingleStageFile(repo, path, **kwargs)
+        return make_dvcfile(repo, path, **kwargs)
+
+
+DVCFile = Union["PipelineFile", "SingleStageFile"]
+
+
+def make_dvcfile(repo: "Repo", path: AnyPath, **kwargs: Any) -> DVCFile:
+    _, ext = os.path.splitext(str(path))
```
Comment (skshetry) [inline dvc/dvcfile.py:400]: We need to fix the `Path` for typing. For now, hacked to use `str()`.

File: dvc/repo/run.py [hunk 3]
```diff
@@ -1,82 +1,29 @@
...
 @scm_context
-def run(self, fname=None, no_exec=False, single_stage=False, **kwargs):
-    from dvc.dvcfile import PIPELINE_FILE, Dvcfile
-    from dvc.exceptions import InvalidArgumentError, OutputDuplicationError
-    from dvc.stage import PipelineStage, Stage, create_stage, restore_meta
-    from dvc.stage.exceptions import InvalidStageName
+def run(
+    self: "Repo",
+    fname: str = None,
+    no_exec=False,
```
Comment (skshetry) [inline dvc/repo/run.py:17]:
```suggestion
    no_exec: bool = False,
```

File: dvc/stage/utils.py [hunk 4]
```diff
@@ -251,3 +262,98 @@ def split_params_deps(stage):
...
+    params = chunk_dict(parse_params(kwargs.pop("params", [])))
+    stage = create_stage(
+        stage_cls, repo=repo, path=path, params=params, **kwargs
+    )
+    restore_meta(stage)
+
+    try:
+        if kwargs.get("force", True):
+            with suppress(ValueError):
+                repo.stages.remove(stage)
```
Comment (skshetry) [inline dvc/stage/utils.py:351]: Thinking of keeping `stages` in a dictionary as `repo.stage_d` and maintain compatibility by returning `stage_d.values()` as a `repo.stages`.

File: dvc/repo/run.py [hunk 5]
```diff
@@ -91,32 +38,13 @@ def run(self, fname=None, no_exec=False, single_stage=False, **kwargs):
...
-    try:
-        if kwargs.get("force", True):
-            with suppress(ValueError):
-                self.stages.remove(stage)
-        else:
-            _check_stage_exists(dvcfile, stage)
-        self.check_modified_graph([stage])
-    except OutputDuplicationError as exc:
-        raise OutputDuplicationError(exc.output, set(exc.stages) - {stage})
+    check_graphs(self, stage, force=kwargs.get("force", True))
```
Comment (skshetry) [inline dvc/repo/run.py:47]: I don't feel good about this, but the `can_be_skipped` checks make it difficult to keep it within `create_stage_from_cli`.

Comment (efiop) [reply dvc/repo/run.py:47]: @skshetry Yeah, was thinking about dropping that logic in 2.0 too.

**Review by efiop — APPROVED** [review]

**Pull request merged**